LEAN-3961: consolidate section category definitions

mbetamony opened the pull request:
1- defined a new type for the [section-categories](https://github.com/Atypon-OpenSource/manuscripts-transform/pull/158/files#diff-bd504b993b890189fe863472a904b264c51a45cca0a7e7cac699e5f9348be749R152): 
- id
- name
- groupIDs
- isDisabled
- synonyms -> includes the different names we used in [chooseSecCategoryByType](https://github.com/Atypon-OpenSource/manuscripts-transform/blob/1467dccadced5a198f68ca31de18f31a17477a6a/src/transformer/section-category.ts#L246C14-L246C41) and [chooseSecCategoryByTitle](https://github.com/Atypon-OpenSource/manuscripts-transform/blob/1467dccadced5a198f68ca31de18f31a17477a6a/src/transformer/section-category.ts#L350)
- titles -> used for autocomplete 
- isEditable -> used in [body-editor](https://github.com/Atypon-OpenSource/manuscripts-body-editor/blob/939371c0b65077c4a2402dc33e618c0c8963aff5/src/lib/section-categories.ts#L32)

I removed all properties that weren't used.

2- renamed some sections so we don't need to handle specific cases that was done by [chooseSecType](https://github.com/Atypon-OpenSource/manuscripts-transform/blob/1467dccadced5a198f68ca31de18f31a17477a6a/src/transformer/section-category.ts#L202)

3- added the type [ManuscriptTemplate](https://github.com/Atypon-OpenSource/manuscripts-transform/pull/158/files#diff-bd504b993b890189fe863472a904b264c51a45cca0a7e7cac699e5f9348be749R162) in the transformer to support having section categories inside the template (we can probably get rid of `createdAt` and `updatedAt`, we kept those after removing manuscripts-data, not sure why)

4- this doesn't belong here but i removed [table-cell-styles.ts](https://github.com/Atypon-OpenSource/manuscripts-transform/pull/158/files#diff-cdfb14b87d6d9db9de31251d8a302378adca060cb6c2f90df16392f61988160b), it doesn't seem to be used anywhere

mnatsheh commented on src/schema/types.ts:
```diff
@@ -147,3 +148,24 @@ export type DataTrackedAttrs = {
+  isEditable: boolean
+  isDisabled?: boolean
+  titles?: string[]
+  groupIDs?: string[]
```
The idea with `groupIDs` never materialized, so I think we should revisit it. Maybe we can get rid of it completely, and update the template to have a separate list for body section categories and backmatter section categories?

mbetamony replied:
So, you mean we'd have 4 lists? abstracts, body, backmatter and other sections? And in the [body-editor](https://github.com/Atypon-OpenSource/manuscripts-body-editor/blob/6d24b2031be40663f75b20d2f03d089ca3ac5582/src/plugins/section-category/section-category-utils.ts#L209) we'd check if the section belongs to a certain list?

mbetamony replied:
It also seems to me that the purpose of groupIDs being an array is lost, the usages assume that we can only have one group id for a section. Can we not have a section that can have multiple group ids?

mnatsheh commented on src/schema/types.ts:
```diff
@@ -147,3 +148,24 @@ export type DataTrackedAttrs = {
+  synonyms: string[]
+  isEditable: boolean
+  isDisabled?: boolean
+  titles?: string[]
```
Can we really have a category without a title?

mbetamony replied:
I'm using the titles property for auto-complete (which only a few sections use). I had to differentiate between titles and synonyms because the old titles we had weren't the ones used by the autocomplete but were used for the import and I didn't think we want to support all previous titles for autocomplete.

mbetamony replied:
If we get rid of the name, then can probably have titles as required and use the first title as the name.

mnatsheh commented on src/schema/types.ts:
```diff
@@ -147,3 +148,24 @@ export type DataTrackedAttrs = {
+
+export type SectionCategory = {
+  _id: string
+  name: string
```
I don't think this is needed, the "titles" should be enough.

mbetamony replied:
Removed.

mnatsheh commented on src/jats/exporter/jats-exporter.ts:
```diff
@@ -2118,7 +2114,10 @@ export class JATSExporter {
       const fnType = fn.getAttribute('fn-type')
       if (fnType) {
-        fn.setAttribute('fn-type', chooseJatsFnType(fnType))
+        fn.setAttribute(
```
what's the purpose of this? why is the "fn-type" not correct from the start?

mbetamony commented:
It seems that even though we have an `abstracts` node, the way its handled is that we have an abstract node that will have an abstracts section (section node) inside it, so we might want to need to keep the 'abstracts' category (though this might require us to have the 'abstracts' back as a section group). Other thing might be we change the way we import the abstracts; it seems that we expect different types of abstract sections, and we add them to the abstract node (check [moveAbstracts](https://github.com/Atypon-OpenSource/manuscripts-transform/blob/b99ab1a7f48dc941f385fd3401af18ecb6ccc205/src/jats/importer/jats-transformations.ts#L129))

mnatsheh commented:
> Further, the History tab provides access to Snapshots, so in those other apps where the Editor is used, if I'm a Viewer would the approach suggested here mean that I'd sometimes have access to snapshots (if current assignee had made a change) and sometimes not have access to snapshots (if current assignee has not yet made a change... or they have made changes, then accepted them)?


This is a transient operation though, we still parse that into an abstracts node in the end.

mnatsheh commented:
> And about changing the templates to have the csl id instead of the bundle, I'm a bit hesitant since this won't be backwards compatible with the WM, we'd probably need to update it, let me know your thoughts.

That's fine let's keep it for now.

mnatsheh commented on src/lib/sectionCategories.ts:
```diff
@@ -0,0 +1,25 @@
+/*!
```
Let's call this `section-categories.ts`.

mbetamony replied:
Done.

mnatsheh commented on src/schema/migration/migration-scripts/3.0.12.ts:
```diff
@@ -0,0 +1,47 @@
+import { JSONNode } from '../migrate'
+import { MigrationScript } from '../migration-script'
+
+export class Migration3010 implements MigrationScript {
```
You need to add this to the list in `index.ts` for it to take effect I think.

mnatsheh commented on src/schema/types.ts:
```diff
@@ -147,3 +148,27 @@ export type DataTrackedAttrs = {
   createdAt: number
 }
+
+export enum SectionGroup {
```
It might be simpler to just use a union type:

`export type SectionGroup = 'body' | 'backmatter'`

mbetamony replied:
Done.

mbetamony commented on src/jats/__tests__/__snapshots__/jats-importer.test.ts.snap:
```diff
@@ -796,7 +796,7 @@ Object {
       "content": Array [
         Object {
           "attrs": Object {
-            "category": "MPSectionCategory:abstract",
+            "category": "",
             "dataTracked": null,
             "id": "MPSection:test",
           },
```
@mnatsheh, this is what I meant by my [comment](https://github.com/Atypon-OpenSource/manuscripts-transform/pull/158#issuecomment-2450539295), the sections under the `abstracts` node are added into a section called abstract. for example, this is what the dom-parser receives from [jats-import.xml](https://github.com/Atypon-OpenSource/manuscripts-transform/blob/master/src/jats/__tests__/__fixtures__/jats-import.xml)
```
<sec sec-type="abstracts"><title>Abstracts</title>
  <sec sec-type="abstract"><title>Abstract</title>
     *** other sections, e.g Objectives
```
I'm not sure if that behavior is intended or we need to fix it.

mbetamony commented on src/jats/__tests__/__snapshots__/jats-importer.test.ts.snap:
```diff
@@ -1284,7 +1284,7 @@ Object {
         },
         Object {
           "attrs": Object {
-            "category": "MPSectionCategory:materials-method",
```
Are we sure we're not importing this section category anymore?

mnatsheh commented on src/jats/__tests__/data/section-categories.ts:
```diff
@@ -0,0 +1,227 @@
+    isUnique: false,
+  },
+  {
+    id: 'abstract-graphical',
```
We should not need this since this has its own node.

mbetamony replied:
Removed.

mbetamony commented on src/jats/__tests__/section-categories.json:
```diff
@@ -0,0 +1,266 @@
+[
+  {
+    "_id": "MPSectionCategory:abbreviations",
```
This is done.

mbetamony commented on src/schema/types.ts:
```diff
@@ -147,3 +148,24 @@ export type DataTrackedAttrs = {
 }
+
+export type SectionCategory = {
+  _id: string
```
Changed.

mbetamony commented:
The reason for changing the name of the `introduction` section to `intro` is because we export it as `intro`. (Same for other sections in this [function](https://github.com/Atypon-OpenSource/manuscripts-transform/blob/1467dccadced5a198f68ca31de18f31a17477a6a/src/transformer/section-category.ts#L202))

mbartenev-atypon commented on src/__tests__/data/project-dump-2.json:
were we not using any of those?

mnatsheh replied:
These are json-schema files, which we should not be using.

mnatsheh commented on src/jats/__tests__/data/section-categories.ts:
```diff
@@ -0,0 +1,220 @@
+    group: 'body',
+    isUnique: false,
+  },
+  {
+    id: 'section',
+    titles: ['Section'],
+    synonyms: [],
+    group: 'body',
+    isUnique: false,
+  },
+  {
+    id: 'subsection',
+    titles: ['Subsection'],
+    synonyms: ['subsection'],
+    group: 'body',
+    isUnique: false,
+  },
```
We can get rid of these now.

mnatsheh commented on src/lib/section-categories.ts:
```diff
@@ -14,24 +14,12 @@
-  _id: SectionGroupTypeID
-  title: string
-}
+export const isSubsection = (categoryID: string) => categoryID === 'subsection'
```
We can get rid of this.

mnatsheh reviewed (commented):
Per discussion with Fahid, we can get rid of the "subsection" category and its usages. We should also update the dom-parser to no inject that sec-type when importing.

mbetamony commented:
> Per discussion with Fahid, we can get rid of the "subsection" category and its usages. We should also update the dom-parser to no inject that sec-type when importing.

Removed, and I updated the migration to remove the sec-type `subsection`.
